lang/rust/rust.py

```diff
@@ -225,11 +234,141 @@ def code_comment_documentation_block_inner():
+
+    To start a tuple type, say "tuple"; to end a tuple type, say "finish".
+    If you don't end a tuple type, it'll automatically be closed.
+
+    For example:
+    'vector' -> 'Vec<>'
+    'vector eye thirty two' -> 'Vec<i32>'
+    'hash map borrowed mute eye thirty two string' -> 'HashMap<&mut i32, String>'
+    'borrowed mute hash map eye thirty two string' -> '&mut HashMap<i32, String>'
+    'tuple eye thirty two character string' -> '(i32, char, String)'
+    'hash map tuple boolean boolean finish string' -> 'HashMap<(boolean, boolean), String)'
+    """
+
+    def parse(tokens: Iterator[Tuple[str, int, str]]) -> Iterator[str]:
+        """Consume some number of tokens, yielding a sequence of strs That can be joined with '' to give a single type."""
+        try:
+            start, expected_args, end = next(tokens)
+        except StopIteration:
+            # this seems silly but is mandated by PEP 479
+            return
+
+        if start == FINISH:
+            # we haven't started a tuple, and this isn't a real type, so suppress it
+            return
+
+        if expected_args == 0:
+            # simplest case, a non-generic type
+            yield start
+
+        elif expected_args > 0:
+            # a generic type with a concrete number of arguments
+            yield start
+            for i in range(expected_args):
+                # each `yield from` corresponds to a single type.
+                # if we run out of tokens, this won't throw an exception, it'll just do nothing--
+                # which is fine, we still want to insert commas and brangles
+                yield from parse(tokens)
+
+                if i < expected_args - 1:
+                    yield ", "
+            yield end
+        else:
+            # expected_args < 0, ie an n-ary object like a tuple
+            yield start
+            written = 0
+            while True:
+                # peek
+                try:
+                    next_start, next_expected, next_end = next(tokens)
+                    stopping = next_start == FINISH
+                except StopIteration:
+                    # stream has ended, but we're inside an n-ary:
+                    # optimistically close it.
+                    stopping = True
+
+                if stopping:
+                    if written == 1:
+                        # (T, ) != (T), so we can't get rid of the comma in this case
+                        yield ", "
+                    break
+                # we're not done, but we might need a comma
+                elif written > 0:
+                    yield ", "
+
+                # reconstruct the iterator and do this step
+                yield from parse(chain([(next_start, next_expected, next_end)], tokens))
+                written += 1
+            yield end
+
+    return "".join(parse(iter(m)))
+
+
+def code_type_(*args):
+    return code_type([generic_type(arg) for arg in args])
+
+
+assert code_type_(["Vec"]) == "Vec<>"
```

Comment:
Ideally these asserts would go in a pytest `test_foo.py` file, so that failures here don't break a language support entirely.

Reply:
👍